core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -513,89 +513,122 @@ private[spark] class Master(
   }
 
   /**
-   * Can an app use the given worker? True if the worker has enough memory and we haven't already
-   * launched an executor for the app on it (right now the standalone backend doesn't like having
-   * two executors on the same worker).
+   * Can an app use the given worker?
    */
-  def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+  private def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
+    val enoughResources = worker.memoryFree >= app.desc.memoryPerExecutorMB && worker.coresFree > 0
+    val allowToExecute = app.desc.maxCorePerExecutor.isDefined || !worker.hasExecutor(app)
+    allowToExecute && enoughResources
```

Comment:
I'm not sure if we need `allowToExecute` here. If `maxCorePerExecutor` is not defined then we will grab all the cores, in which case `worker.coresFree` will be 0.